### PR TITLE
Do not add BOM when commitEncoding is set to upcase UTF-8

### PR DESCRIPTION
**Pull request by orgads**
commitEncoding is case-insensitive, and the predefined UTF8 encoding is using lowercase key utf-8. Convert to lowercase to guarantee that it matches.

## Test methodology 
- `git config i18n.commitEncoding UTF-8`
- Commit using Git Extensions
- `git show` shows a leading space before the commit subject.

## Test environment(s) 

- GIT 2.38.1
- Windows 10

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitCommands/Settings/ConfigFileSettings.cs [hunk 1]
```diff
@@ -136,7 +136,7 @@ public Encoding FilesEncoding
 
         private Encoding? GetEncoding(string settingName)
         {
-            string encodingName = GetValue(settingName);
+            string encodingName = GetValue(settingName).ToLower();
```
Comment (RussKie) [inline GitCommands/Settings/ConfigFileSettings.cs:139]: Could you please elaborate how this is addressing the issue?

Comment (orgads) [reply GitCommands/Settings/ConfigFileSettings.cs:139]: I wrote it in the commit message.

Suppose the setting is UTF-8. Here it is converted to lower case, so `encodingName` is `utf-8`. This value exists in `AppSettings.AvailableEncodings`, and it is mapped to BOM-less UTF-8.

**Comment by mstv** [comment]
Have you tried to configure the encodings which shall be used in GE? I would expect that the encodings will match exactly then.

grafik

grafik

**Comment by orgads** [comment]
Yes, this works, but if the encoding was configured *by command line* as upcase, it doesn't match.

Do you have a reason*not* to have this extra layer of protection?

File: GitCommands/Settings/ConfigFileSettings.cs (hunk 1 above)
Comment (RussKie) [reply GitCommands/Settings/ConfigFileSettings.cs:139]: Thank you for the clarification. Please add a comment to the code explaining this, so that this change won't get undone by accident.

Comment (RussKie) [inline GitCommands/Settings/ConfigFileSettings.cs:139]:
```suggestion
            string encodingName = GetValue(settingName).ToLowerInvariant();
```

**Review by RussKie — COMMENTED** [review]
It would be great to add few tests to verify the issue is fixed and won't regress in the future.
A test would create a test repo, configure `i18n.commitEncoding` with upper/lower/mixed case and verify the encoding is expected.

We have few similar tests, e.g., https://github.com/gitextensions/gitextensions/blob/master/UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs

File: GitCommands/Settings/ConfigFileSettings.cs (hunk 1 above)
Comment (orgads) [reply GitCommands/Settings/ConfigFileSettings.cs]: Done

**Comment by orgads** [comment]
> It would be great to add few tests to verify the issue is fixed and won't regress in the future. A test would create a test repo, configure `i18n.commitEncoding` with upper/lower/mixed case and verify the encoding is expected.
> 
> We have few similar tests, e.g., https://github.com/gitextensions/gitextensions/blob/master/UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs

Done. Not sure if it's the right place for them though :)

File: GitCommands/Settings/ConfigFileSettings.cs (hunk 1 above)
Comment (orgads) [reply GitCommands/Settings/ConfigFileSettings.cs]: Done

File: UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs [hunk 2]
```diff
@@ -392,6 +392,17 @@ public void ResetCommitMessage()
...
+        [Test, TestCaseSource(nameof(EncodingNames))]
+        public void MergeOrCommitMessage_should_use_correct_encoding_when_name_is_upcase(
+            string encodingName)
+        {
+            AppSettings.LoadSettings();
+            _referenceRepository.Module.EffectiveConfigFile.SetString("i18n.commitEncoding", encodingName);
+            _referenceRepository.Module.CommitEncoding.Preamble.Length.Should().Be(0);
+        }
+
+        private static IEnumerable<string> EncodingNames => new List<string> { "utf-8", "Utf-8", "UTF-8" };
```
Comment (gerhardol) [inline UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs:404]: Add to the beginning of the file.

**Review by gerhardol — COMMENTED** [review]
Maybe better in GitModuleTests.cs

File: UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs (hunk 2 above)
Comment (orgads) [reply UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs:404]: Done

File: UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@ public class CommitMessageManagerTests
         private ReferenceRepository _referenceRepository;
 
         private readonly string _workingDirGitDir = @"c:\dev\repo\.git";
-        private readonly Encoding _encoding = Encoding.UTF8;
+        private readonly Encoding _encoding = new UTF8Encoding(false);
```
Comment (gerhardol) [inline UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs:24]: What is the motivation for this change?

**Review by gerhardol — COMMENTED** [review]
+1
Someone else to approve too.

File: UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs (hunk 3 above)
Comment (orgads) [reply UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs:24]: To disable BOM also for this test. It's not mandatory, but since I already handled similar cases, I preferred to have no BOM at all.

**Review by vbjay — CHANGES_REQUESTED** [review]
I would recommend adding a test that proves that changing case of config actually results in BOM or not. I'd also verify files are not affected. Add 1 file that has BOM and one that does not and commit. Then delete the working directory files and reset hard. Verify BOM in the one file.

**Comment by orgads** [comment]
> I would recommend adding a test that proves that changing case of config actually results in BOM or not. I'd also verify files are not affected. Add 1 file that has BOM and one that does not and commit. Then delete the working directory files and reset hard. Verify BOM in the one file.

This is the commit message, not a file in the repo.

I added another test for the message content.

**Comment by vbjay** [comment]
> > I would recommend adding a test that proves that changing case of config actually results in BOM or not. I'd also verify files are not affected. Add 1 file that has BOM and one that does not and commit. Then delete the working directory files and reset hard. Verify BOM in the one file.
> 
> This is the commit message, not a file in the repo.
> 
> I added another test for the message content.

Correct. I am suggesting a test that proves files are not affected. A red herring test to spot issues with file content being altered.

**Comment by orgads** [comment]
Why? How is this related to the commit message authoring? How would this test look like?

**Comment by vbjay** [comment]
Create a test repo and create 2 files. 

- file with BOM
- file without BOM

Commit these files under your different cases. Remove files and then checkout each commit files to verify the files still are one BOM and one not. 

The test would confirm that changing the setting doesn't affect the files. I expect it to be fine but I'd like proof that whatever case of i18n.commitEncoding does not affect file content and only alters commit message.  This woukd more detect git changed behaviour but knowing what the setting does and doesn't cover and verifying it hasn't changed on us would be good.

**Comment by orgads** [comment]
Thank you.

**Comment by orgads** [comment]
I'll merge my tests, no reason to have 2 of them.

File: UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs (hunk 3 above)
Comment (RussKie) [reply UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs:24]: I suggest adding a comment to spare others from searching this in docs.

**Review by RussKie — APPROVED** [review]
👍

File: UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs (hunk 3 above)
Comment (mstv) [reply UnitTests/GitCommands.Tests/CommitMessageManagerTests.cs:24]:
```suggestion
        private readonly Encoding _encoding = new(encoderShouldEmitUTF8Identifier: false);
```
makes it clear.

**Review by mstv — COMMENTED** [review]
LGTM

**Comment by orgads** [comment]
Done

**Comment by mstv** [comment]
Could you add your sign-off to `Contributors.txt` yet, please?

**Comment by orgads** [comment]
Done

**Review by mstv — APPROVED** [review]
Thank you
